Upgrades for Generators

cholmcc opened the pull request:
This superseeds [this merge request](https://github.com/AliceO2Group/AliceO2/pull/11913) 

The text of that merge request: 

Please see the commit logs of the individual commits.

**TL;DR**: 

- `MCEventHeader`: Pre-defined keys for information
- `GeneratorPythia8`: Export _all_ available information to `MCEventHeader` 
- `GeneratorHepMC`:
  - Export _all_ available information to `MCEventHeader` 
  - Read events from external, spawned background program via FIFO 
- `GeneratorTParticle`: 
  - Reads from `TChain` with `TClonesArray` branch with `TParticle` objects 
    - Can read from existing files, _or_
    - from file produced by external, spawned background program

The changes to `GeneratorHepMC` allows us to use _any_ event generator that can write HepMC events. 

Examples of use are given in log messages and code comments. 

Yours,

_Christian_

cholmcc committed:
Introduce pre-defined keys for MC event header info

A number of keys into the ME event header information
mapping is defined.   This is to ensure that code will
use the same keys when ever information is set.

Additional, non-predefined keys, are still possible.

This makes it much more robust when we ask for specific
MC information from the event header, such as

- cross-section(s)
- weight(s)
- Heavy-ion "geometry" parameters
  - Npart in projectile and target
  - Ncoll in various views
    - Overall
    - Hard
    - wounded-nucleon on nucleon
    - nucleon on wounded-nucleon
    - wounded on wounded
- Parton distribution function parameters

This is crucial for building a HepMC event structure which
can be passed on to say for example Rivet analyses.

cholmcc committed:
Export _full_ header information to MC event header

The generator has been changed so that it exports
_all_ relevant and available information from Pythia to
the MC event header, including heavy-ion "geometry"
parameters.  In particular, the information is stored
in an HepMC compatible way for later user by f.ex.
Rivet.

Note, the current code counts up the number of collisions
by it self.  However, the authors of Pythia have another
way of doing that.   The code is now there to do it the
same way as the Pythia authors, but is currenly disabled.

We should decide which is the appropriate way to count
Ncoll.  I would recommend to follow how the Pythia
authors do it.

cholmcc committed:
Full header read-in and external program

This change does two things:

**Full header**

_All_ information available in the HepMC event header is
propagated to the MC event header information map.  This
includes

- Heavy-ion "geometry" parameters (b,Ncoll,Npart,...)
- Cross-section(s)
- Weight(s)
- PDF information
- and other attributes defined

This is so that we can build a full HepMC event structure later -
for example to pass to Rivet analyses

** External program **

The functionality of the generatator is expanded so that it may
spawn an event generator program, say `eg`.

- The generator opens a FIFO
- The generator then executes the program `eg` in the background
  - The `eg` program is assumed to write HepMC event records on
    standard output, which is then redirected to the FIFO
- The generator reads events from the FIFO

For this to work, a number of conditions _must_ be met by the
`eg` program:

- It _must_ write events in the HepMC event format
- It _must_ write the HepMC event records to standard output
- It _cannot_ write anything else but the HepMC event record to
  standard output
- It _must_ accept the command line option `-n NEVENTS` to
  set the number of events to generate.

If a particular `eg` program does not meet these requirements, then
a simple shell script can be defined to wrap the `eg` appropriately.
For example, the CRMC program `crmc` _can_ write HepMC events to
standard output, but it will also dump other stuff there.  Thus,
we can provide the script

    #!/bin/sh

    crmc $@ -o hepmc3 -f /dev/stdout | \
       sed -n 's/^\(HepMC::\|[EAUWVP] \)/\1/p'

which simply filters the output of `crmc`.  Another EG program
may not accept the `-n EVENTS` command line option, but rather has
the command line option `--nevents`, so then we would do something
like

    #!/bin/sh
    cmdline="eg-program -o /dev/stdout "

    while test $# -gt 0 ; do
       case x$1 in
       x-n) cmdline="$cmdline -n $2"; shift ;;
       *)   cmdline="$cmdline $1" ;;
       esac

       shift
    done

    $cmdline

The command line to run is specified as

    --configKeyValues "HepMC.progCmd=<program and options>"

and can include not only the program name but also other
options to the program.  For example

    --configKeyValues "HepMC.progCmd=crmc -m 5 -i 20800820 -I 20800820"

for Pb-Pb collisions with Hijing.

With this change, we can use _any_ event generator which is capable to
write out its event records in the HepMC format.

cholmcc committed:
New generator GeneratorTParticle

The generator `GeneratorTParticle` will read in particles
from a `TChain` containing a branch with a `TClonesArray` of
`TParticle` objects.

The generator can operate in two modes

- Data is read from a file(s)
- Data is read from a file being generated by a child
  program

The first mode is selected by

    -g tparticle --configKeyValues "TParticle.fileNames=foo.root,bar.root"

The second mode is selected by

    -g tparticle --configKeyValues "TParticle.progCmd=<program and options>"

For this latter mode, see also recent commit to `GeneratorHepMC`

Above `<program and options>` specifiy a program to spawn in the
background which will write to a specified file (temporary file).
Suppose the program is called `eg`, then the following _must_ be
possible

    eg -n NEVENTS -o OUTPUT_FILENAME

That is, `eg` _must_ accept the option `-n` to set the number of
events to produce, and the option `-o` to set the output file name
(a ROOT file).

The name of the `TTree` object in the file(s) can be set with

    --configKeyValues "TParticle.treeName=<name>"

(defaults to `T`), and similar for the branch that contains the
`TClonesArray` of `TParticle`

    --configKeyValues "TParticle.branchName=<name>"

(defaults to `Particles`).

The generator `GeneratorTParticle` _does not_ import any header
information into the simulation event record.   Some proper
convention could be decided upon, e.g., one that tracks the
HepMC event record format.

alibuild committed:
Please consider the following formatting changes

Commit:
Merge pull request #1 from alibuild/alibot-cleanup-11913

Please consider the following formatting changes to #11913

cholmcc committed:
Fixes for copyright stuff - sigh!

cholmcc committed:
One more comment

alibuild committed:
Please consider the following formatting changes

Commit:
Merge pull request #2 from alibuild/alibot-cleanup-11913

Please consider the following formatting changes to #11913

cholmcc committed:
Whitespace

cholmcc committed:
untabify

Commit:
Merge branch 'AliceO2Group:dev' into cholmcc_generatos

cholmcc committed:
Added documentation

cholmcc committed:
Refactoring to HepMC and TParticle Generators

The classes `GeneratorHepMC` and `GeneratorTParticle` is refactored to
derive from the (second) base class `GeneratorFileOrCmd`.

`GeneratorFileOrCmd` provides common infrastructure to specify

- File(s) to read events from (ROOT files in case of
  `GeneratorTParticle`, and HepMC files in case of `GeneratorHepMC`),
  _or_
- Which commmand to execute and with which options.

It also provides infrastructure to make unique temporary names, a FIFO
to read from (child program writes to), and so on.

These are all configured through configuration keys prefixed by
`FileOrCmd.`.

Other changes include

- `GeneratorHepMC` will open _any_ file that HepMC supports - ASCII,
  compressed ASCII, HEPEVT, etc.
- Through the use of `GeneratorFileOrCmd` the command line option flags
  for specifying seed (default: `-s`), number of events (default `-n`),
  largest impact parameter (defautl: `-b`), output (default: `>`), and
  so on can be configured via configuration key values
- `GeneratorHepMC` and `GeneratorTParticle` are passed the
  `GeneratorFileOrCmdParam` as well as specific `GeneratorHepMCParam`
  and `GeneratorTParticleParam`, respectively, objects by
  `GeneratorFactor` and sets the internal parameters accordingly. This
  hides the specifics of the parameters from `GeneratorFactory`.

cholmcc committed:
Changed key prefix

cholmcc committed:
Add new sub-dirs

cholmcc committed:
Merged changes from previous branch

cholmcc committed:
Undo adding AODToHepMC

cholmcc committed:
Undo bad commit - these changes are needed but should be in different MR

cholmcc committed:
Formatting fixes

cholmcc committed:
Renamed _script_ (not macro) to end in .macro (sigh)

cholmcc committed:
Renamed _script_ (not macro) to end in .macro (sigh)

cholmcc committed:
Merge branch 'cholmcc_generatos' into cholmcc_generators_more

cholmcc committed:
Fix formatting

cholmcc committed:
Fix formatting

cholmcc committed:
Merge branch 'cholmcc_generatos' into cholmcc_generators_more

cholmcc committed:
Fix formatting

cholmcc committed:
Merge branch 'cholmcc_generatos' into cholmcc_generators_more

cholmcc committed:
Fix formatting - that checker is ridiculously pedantic about things and makes life so much more difficult than it needs to be for very little gain

cholmcc committed:
Fix formatting

cholmcc committed:
Fix formatting - that checker is ridiculously pedantic about things and makes life so much more difficult than it needs to be for very little gain

cholmcc committed:
Merge branch 'cholmcc_generatos' into cholmcc_generators_more

cholmcc committed:
Stupid annoying formatting checker - it is really f**cking ridiculous that it is so pedantic that you cannot have two blank lines at the end of a simple shell script - this is studpid and _very_ counter productive.  Please please _please_ change the policy on this

cholmcc committed:
Undo last faulty commit and fix formatting

This is exactly what happens when you make the stupid arse, over zealous
code checkers.  I get frustrated and out of sheer oversight, commit some
stuff that shouldn't be committed and then I have to go back and fixt it
again.  That is so annoying, counter-productive, and down right silly.

This could all be avoided if the code checker did _reasonable_ checks.
Who cares if there's an extra blank line in a shell script, or other
such annoying things.

In fact, if it is so important, then git hooks that _automatically_
fixes these issues should be enstated instead of relying on the
developers to do this f**ked up things over and over again.

Please, please, _please_ fix this.

cholmcc committed:
Merge branch 'cholmcc_generatos' into cholmcc_generators_more

cholmcc committed:
Fix whitespace - these changes is so messed up

ktf commented on run/SimExamples/HepMC/child.sh:
```diff
@@ -0,0 +1,51 @@
+#!/usr/bin/bash
```
```suggestion
#!/usr/bin/env bash
```

Commit:
Update run/SimExamples/HepMC/child.sh

Co-authored-by: Giulio Eulisse <10544+ktf@users.noreply.github.com>

sawenzel commented on Generators/src/GeneratorFileOrCmd.cxx:
```diff
@@ -0,0 +1,207 @@
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+/// \author R+Preghenella - August 2017
```
This is probably a copy-paste error!?

cholmcc replied:
You mean the `\author` - yes, that's copy'n'paste oversight.  I don't mind if you don't mind.

ktf replied:
```suggestion
```

ktf replied:
We should avoid any authorship information beyond the license, actually.

ktf committed:
Update Generators/src/GeneratorFileOrCmd.cxx

cholmcc commented:
Failed test comments: 

- `build/AliceO2/O2/o2/macOS` - from `curl` - not part of the MR, false negative 
- `build/AliceO2/O2/o2/macOS-arm` - Test `o2sim_hepmc` uses wrong key `HepMC.fileName` - should be `FileOrCmd.fileNames=`.  Will fix this and commit 

Whoops, @ktf did a commit so I cannot see the rest of the errors - will have to wait I guess.

ktf commented:
Just click on the red cross on your commit, it will show the details for the checks.

alibuild commented:
Error while checking build/O2/fullCI for fba7f8c01aa8a36e6a190a5fa6dbbbb671fdce36 at 2023-10-09 13:27:
```
## sw/BUILD/O2-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'


## sw/BUILD/QualityControl-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'


## sw/BUILD/O2Physics-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[0 more errors; see full log]
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/12032/fba7f8c01aa8a36e6a190a5fa6dbbbb671fdce36/build_O2_fullCI/pretty.html).

sawenzel commented on Generators/include/Generators/GeneratorHepMCParam.h:
```diff
@@ -30,8 +30,7 @@ namespace eventgen
 
 struct GeneratorHepMCParam : public o2::conf::ConfigurableParamHelper<GeneratorHepMCParam> {
-  std::string fileName = "";
-  int version = 2;
```
We have a problem here: `HepMC.filename` is used in existing examples and tests. The CI fails since you take this away. Also, there seems to be an inconsistency in the description above: In one point you say that `.filename` is no longer supported ... but then `filename ` occurs in newly added examples. It would be best to avoid changing things too much...  so better keep the existing keys and functionality. I would also like to encourage a review from @preghenella as original author on this.

cholmcc replied:
Hi Sandro, 

Yes, the newly added documentation is a bit out of sync - but note the **In the near future** section.   This was because I did not know which MR you would take.    I've added code to `GeneratorHepMC` which will throw a `fatal` if the old key is used, and gives an appropriate message about what to do instead. 

Note that the previous code only allowed one to read from 1 file, and that you had to know the version (and specify it) of the format.   The new code allows for multiple input files and the it figures out the format (including version) itself.   It also supports all the file formats supported by the HepMC3 install (ROOT, LHEF, HepEVT, remote access ROOT, etc. are some of the options in addition to plain ASCII HepMC). 

I can update the docs and add them to the MR.

cholmcc replied:
Fixed docs and other places where `HepMC.fileName` was used.

sawenzel reviewed (changes requested):
see comment about `HepMC.filename`

cholmcc commented:
Continuing the error analysis: 

- `build/O2/fullCI` pedantic errors about braces for single-statement `if`, `while`, etc.  Really not a very useful or meaningful thing.  The rest are all false-positives wrt to this MR.  I will commit fixes. 
- `build/O2/o2` - same problem as `build/AliceO2/O2/o2/macOS-arm` - see previous message 
- `build/O2/o2-cs8` - as above. 
- `build/O2/o2-cs8-perftest` - as above
-

cholmcc committed:
Fix test case and other similar fixes

- The test case `o2sim-hepmc` fixed to use proper config key
- `GeneratorHepMC` will give error if old config key used
- `GeneratorHepMC` will give warning if version key is set - the code
  deduces the version on its own now.
- Added superflous `{...}` around single statement `if`, `while`, ... -
  boy those checks are silly and counter productive.

cholmcc committed:
Formatting fixes and bring docs into Sync

cholmcc committed:
Fixed example for new keys

cholmcc committed:
Fix whitespace - not even my edit

cholmcc commented:
> see comment about `HepMC.filename`

Fixed - please remove _change request_.

Commit:
Merge branch 'AliceO2Group:dev' into cholmcc_generators_more

cholmcc commented:
Error analysis (If this is not useful, let me know):

`o2-dataflow-cs8-perftest`: 

```
 1/28 Test  #1: o2-qc-test-core .........................***Timeout  30.03 sec
2023-10-10 04:31:39.499645  !  [1;33mWarning - Aggregator name "012345678901234567890" is longer than 16, it might cause name clashes in the DPL workflow[0m
2023-10-10 04:31:39.500666  !  [1;33mWarning - No remote machine was specified for a multinode QC setup. This is fine if running with AliECS, but it will fail in standalone mode.[0m
2023-10-10 04:31:39.500701  !  [1;33mWarning - No remote machine was specified for a multinode QC setup. This is fine if running with AliECS, but it will fail in standalone mode.[0m
2023-10-10 04:31:39.500768  !  [1;33mWarning - too long detector code for a task data origin: XXXXXXXXX, trying to survive with: XXX[0m
2023-10-10 04:31:39.500789  !  [1;33mWarning - too long detector code for a task data origin: XXXXXXXXX, trying to survive with: XXX[0m
2023-10-10 04:31:39.500800  !  [1;33mWarning - too long detector code for a task data origin: XXXXXXXXX, trying to survive with: XXX[0m
2023-10-10 04:31:39.500814  !  [1;33mWarning - too long detector code for a task data origin: XXXXXXXXX, trying to survive with: XXX[0m
2023-10-10 04:31:39.500834  !  [1;33mWarning - PP Task name "SkeletonPostProcessing" is longer than 16, it might cause name clashes in the DPL workflow[0m
2023-10-10 04:31:39.501064     Extracting configuration of a new aggregator MyAggregatorA
2023-10-10 04:31:39.501073        Found a source : MyAggregatorB
2023-10-10 04:31:39.501078           - MyAggregatorB/newQuality
2023-10-10 04:31:39.501084        Found a source : MyAggregatorC
2023-10-10 04:31:39.501088           - MyAggregatorC/newQuality
2023-10-10 04:31:39.501092           - MyAggregatorC/another
2023-10-10 04:31:39.501102     Extracting configuration of a new aggregator MyAggregatorD
2023-10-10 04:31:39.501107        Found a source : MyAggregatorA
2023-10-10 04:31:39.501111           - MyAggregatorA/newQuality
2023-10-10 04:31:39.501115        Found a source : MyAggregatorC
2023-10-10 04:31:39.501119           - MyAggregatorC/newQuality
2023-10-10 04:31:39.501123           - MyAggregatorC/another
2023-10-10 04:31:39.501128     Extracting configuration of a new aggregator MyAggregatorC
2023-10-10 04:31:39.501132        Found a source : dataSizeCheck
2023-10-10 04:31:39.501137           (no QOs specified, we take all)
2023-10-10 04:31:39.501142        Found a source : someNumbersCheck
2023-10-10 04:31:39.501146           (no QOs specified, we take all)
2023-10-10 04:31:39.501172     Extracting configuration of a new aggregator MyAggregatorB
2023-10-10 04:31:39.501181        Found a source : dataSizeCheck1
2023-10-10 04:31:39.501184           (no QOs specified, we take all)
2023-10-10 04:31:39.501188        Found a source : dataSizeCheck2
2023-10-10 04:31:39.501192           - dataSizeCheck2/someNumbersTask/example
2023-10-10 04:31:39.501494 !!! [1;31mError - Could not find the DPL InfoLogger.[0m
2023-10-10 04:31:39.501578 !!! [1;31mError - exception caught and swallowed in templateILDiscardFile : Dynamic exception type: o2::framework::RuntimeErrorRef[0m
2023-10-10 04:31:39.501638  !  [1;33mWarning - Could not get updated config tree in TaskRunner::init() - `qcConfiguration` could not be retrieved[0m
2023-10-10 04:31:39.501653  !  [1;33mWarning - No URL provided for Bookkeeping. Nothing will be stored nor retrieved.[0m
2023-10-10 04:31:39.501666     Loading library libO2QcSkeleton
```

An error in `o2-qc-test-core` which is _unrelated_ to the current MR.  A **false negative**.

- `macOS`:  

```
dyld[31595]: missing symbol called
/Volumes/build/build/alice-ci-workdir/o2/sw/SPECS/osx_x86-64/cgal/4.12.2-local1/cgal.sh: line 10: 31595 Abort trap: 6           curl -kLo cgal.tar.xz "$URL"
```

A problem in `curl` - seems like MacOS couldn't load an appropriate dynamic library - pretty bad, but _not_ related to this MR - a **false negative**

Yours, 

_Christian_

cholmcc commented:
Error analysis:

- `build/AliceO2/O2/o2/macOS-arm` failes due to test `test_Framework_test_RegionInfoCallbackService` which is unrelated to this MR - so a **false negative**

sawenzel reviewed (changes requested):
I had a more thorough look/review into this. Overall the PR is very valuable and a good way forward. However, a couple of things still need attention or careful discussion:

a) My first and foremost worry concerns the introduction backward-incompatible restructuring of how HepMC filenames need to be given by users. `HepMC.filename` is no longer supported. This will need follow-up changes for instance in O2DPG where several PWGs already use this. Is there no way to keep the current `HepMC.filename` key? (you may forward this key to some common internal structure such as `FileOrCmd`)

b) My second comment concerns the newly introduced configurable value `FileOrCmd`. As previously argued with key `TParticle`, I do not find the choice of name appropriate. The key doesn't tell the user what this is for. I understand that this configures how certain external generators (such as HepMC or others) are read/connected. So I would suggest to use something like `GenFileOrCmd`, `GeneratorFileOrCmd` or similar (a name that relates to generator). I can then more easily understand `GenFileOrCmd.filenames`, `GenFileOrCmd.cmd` when I read them (like a namespace).

c) I tried to run the examples in `SimExamples/HepMC` folder. At the moment I was not successful:
   - `crmc` does not have an option `-b`
   - the CRMC that we ship in ALICE with alidist is out of date and does not know `-o hepmc3`
   - even fixing these 2 ... the `child.sh` example does not run for me. It tells `WARNING::ReaderAscii: found unsupported expression in header. Will close the input. HepMC::As EPOS used with FUSION option`. This may depend on the precise crmc commit of course or just need a slightly modified `sed` filter.
 
   It would be good to make sure that the example actually runs fine. I am also trying to setup a second example for STARlight. We should in particular make sure that the mechanism works for HepMC2 and for HepMC3.

sawenzel commented on Generators/include/Generators/GeneratorHepMCParam.h:
```diff
@@ -30,9 +30,9 @@ namespace eventgen
 struct GeneratorHepMCParam : public o2::conf::ConfigurableParamHelper<GeneratorHepMCParam> {
-  std::string fileName = "";
-  int version = 2;
+  int version = 0;
```
You disabled this with the explanation that it can be figured out automatically from the file format.
However, for the FIFO case, a hard-coded HepMC3 reader is instantiated. 

 So either HepMC2 is not supported in this mode (which would be unfortunate) or it would make sense to keep the version field to be able to select the appropriate reader.

cholmcc replied:
Hi Sandro, 

I made a new commit to the MR, which should address your concerns (and some other improvements)

## Back-ward compatibility and Namespace change

- `GeneratorHepMC` accepts the old `HepMC.fileName` configuration
  option (with a deprecation warning)
- `HepMC.version` is honoured when spawning a child process.
  That is, if it is set to 2, then use compatibility reader
  `HepMC3::ReaderAsciiVersion2`, otherwise use the version 3
  reader.

  If option is given with file names (`GeneratorFileOrCmd.fileNames` or deprecated `HepMC3.fileName`), then give warning that it is not used, but that the reader deduces the version it self. 

- Configuration namespace `FileOrCmd` is renamed to
  `GeneratorFileOrCmd`

- Documentation has been updated to reflect these changes

alibuild commented:
Error while checking build/O2/fullCI for 330feea734d57bd000d2135302e47b59c1da1689 at 2023-10-14 02:47:
```
No log files found
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/12032/330feea734d57bd000d2135302e47b59c1da1689/build_O2_fullCI/pretty.html).

cholmcc committed:
Back-ward compatibility and Namespace change

- `GeneratorHepMC` accepts the old `HepMC.fileName` configuration
  option (with a deprecation warning)
- `HepMC.version` is honoured when spawning a child process.
  That is, if it is set to 2, then use compatibility reader
  `HepMC3::ReaderAsciiVersion2`, otherwise use the version 3
  reader.

  If option is given with file names, then give warning that
  it is not used.

- Configuration namespace `FileOrCmd` is renamed to
  `GeneratorFileOrCmd`

- Documentation has been updated to reflect these changes

cholmcc commented:
> Error while checking build/O2/fullCI for [330feea](https://github.com/AliceO2Group/AliceO2/commit/330feea734d57bd000d2135302e47b59c1da1689) at 2023-10-14 02:47:
> 
> ```
> No log files found
> ```
> 
> Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/12032/330feea734d57bd000d2135302e47b59c1da1689/build_O2_fullCI/pretty.html).

Not sure if this failure has anything to do with this MR.  It seems some `git` clones failed 

```
Git command for package 'MLModels' failed.
Command: git fetch -f --tags https://github.com/alisw/MLModels.git +refs/heads/*:refs/heads/*
In directory: /build/nomad/alloc/78727433-50b3-e91c-7f54-7ec8b17f3bdc/ci/local/o2-fullci/sw/MIRROR/mlmodels
Exit code: 128
Git command for package 'capstone' failed.
Command: git fetch -f --tags https://github.com/aquynh/capstone +refs/heads/*:refs/heads/*
In directory: /build/nomad/alloc/78727433-50b3-e91c-7f54-7ec8b17f3bdc/ci/local/o2-fullci/sw/MIRROR/capstone
Exit code: 128
```

I would chalk it up as a **false negative**

cholmcc commented:
> I had a more thorough look/review into this. Overall the PR is very valuable and a good way forward. However, a couple of things still need attention or careful discussion:

OK
 
> a) My first and foremost worry concerns the introduction backward-incompatible restructuring of how HepMC filenames need to be given by users. `HepMC.filename` is no longer supported. This will need follow-up changes for instance in O2DPG where several PWGs already use this. Is there no way to keep the current `HepMC.filename` key? (you may forward this key to some common internal structure such as `FileOrCmd`)

OK, so as you see below, I've made the changes backward compatible - albeit with a deprecation warning.  Hopefully users will upgrade their code and we can remove this "hack" later on down the road. 

> b) My second comment concerns the newly introduced configurable value `FileOrCmd`. As previously argued with key `TParticle`, I do not find the choice of name appropriate. The key doesn't tell the user what this is for. I understand that this configures how certain external generators (such as HepMC or others) are read/connected. So I would suggest to use something like `GenFileOrCmd`, `GeneratorFileOrCmd` or similar (a name that relates to generator). I can then more easily understand `GenFileOrCmd.filenames`, `GenFileOrCmd.cmd` when I read them (like a namespace).

The namespaces for the `Generator...` parameters are all over the places.  Some prefix with `Gen`, some with `Generator`, others, nothing at all.  To spell it out, I changed it to `GeneratorFileOrCmd`.  I don't care too much about this, only that they names may get so long that users may find it a hassle. 

> c) I tried to run the examples in `SimExamples/HepMC` folder. At the moment I was not successful:
> 
> * `crmc` does not have an option `-b`

In the example, I explicitly set the `bMaxSwitch=none` so that the option isn't passed.  Actually a good way of illustrating that feature.  Note, one can specify a max $b$ for CRMC, but I think it is a setting in the `crmc.params` file.  Of course, the `crmc.sh` script could "fix that up", but I didn't want to complicate the example too much. 

> * the CRMC that we ship in ALICE with alidist is out of date and does not know `-o hepmc3`

Well, hopefully that will get upgraded soon :-) 

With the latest change, we could do `-o hepmc` and then pass `HepMC.version=2` as a configuration key-value

If CRMC installed with `alidist` only uses HepMC2, then I'm afraid that it is terribly out of date, and the HepMC events written are not correct.  I believe so because I was the one making the changes for HepMC3 in CRMC and fixed other issues at the same time.  See [here](https://gitlab.iap.kit.edu/AirShowerPhysics/crmc/-/merge_requests/3), [here](https://gitlab.iap.kit.edu/AirShowerPhysics/crmc/-/merge_requests/2), and [here](https://gitlab.iap.kit.edu/AirShowerPhysics/crmc/-/merge_requests/6)

> * even fixing these 2 ... the `child.sh` example does not run for me. It tells `WARNING::ReaderAscii: found unsupported expression in header. Will close the input. HepMC::As EPOS used with FUSION option`. This may depend on the precise crmc commit of course or just need a slightly modified `sed` filter.

That would be fixed by passing `HepMC.version=2`. 

Unfortunately the `HepMC3.deduceReader` (or something like that) will not work with child command, since it is rather aggressive at opening the input file. 

> It would be good to make sure that the example actually runs fine. I am also trying to setup a second example for STARlight. We should in particular make sure that the mechanism works for HepMC2 and for HepMC3.

OK, should work with the latest commit. 

Yours,
_Christian_

cholmcc committed:
Fix for test

- Forgot to change key for test in `run/CMakeLists.txt`
- `crmc.sh` uses `-o hepmc` instead of `-o hepmc3` to accomodate
  older installation of CRMC with `aliBuild`.

cholmcc commented:
Hi all, 

The last commit, I believe, fixes up the request changes.  All tests but [`build/AliceO2/O2/o2/macOS`](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/12032/353068e7b8c8894cf921cc7f66f71246c1977974/build_AliceO2_O2_o2_macOS/pretty.html) succeed, and that test fails due to a deeper MacOS problem launching `curl`.

Yours,
_Christian_

martenole commented:
Hi Christian, Sandro is away this week, but should be back on Monday. The macOS check is broken at the moment, so what concerns the CI your PR is good to go.

alibuild commented:
Error while checking build/O2/fullCI for 353068e7b8c8894cf921cc7f66f71246c1977974 at 2023-10-19 05:11:
```
## sw/BUILD/O2-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'


## sw/BUILD/o2checkcode-latest/log
--
========== List of errors found ==========
++ GRERR=0
++ grep -v clang-diagnostic-error error-log.txt
++ grep ' error:'
/sw/SOURCES/O2/12032-slc8_x86-64/0/Generators/src/GeneratorPythia8.cxx:383:14: error: statement should be inside braces [readability-braces-around-statements]
++ [[ 0 == 0 ]]
++ exit 1
--
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/12032/353068e7b8c8894cf921cc7f66f71246c1977974/build_O2_fullCI/pretty.html).

cholmcc committed:
Fix braces
